### PR TITLE
useSubscription hook

### PR DESCRIPTION
**Pull request by frikille**
### What does this PR do?

Implementing useSubscription in graphql-hooks

This PR is more like a proof of concept at the moment after spending some time learning about different useSubscription implementation.

The current implementation is very simple and is intended to start a discussion on how the final implementation and API should look like. This implementation required the minimal code to get started and it doesn't add new dependencies to the graphql-hooks project.

The GraphQLClient has been modified and it is possible to set a new config option called `subscriptionClient` which must be an instance of `SubscriptionClient` from the (`subscription-transport-ws`)[https://github.com/apollographql/subscriptions-transport-ws] package. By using the `SubscriptionClient` there is no need to implement anything new apart of the `useSubscription` function in `graphql-hooks`. However, I think it's worth the discussion around using the SubscriptionClient from that package or implementing our own as that library doesn't seem to get too much care from its maintainers.

If we decide to implement our own library that probably introduces some new dependencies however that can be mitigated by creating a new package and pass the subscription client instance to the GraphQLClient during instantiation.

### Related issues

#72 

### Checklist

- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md)
- [x] I have added or updated any relevant documentation
- [x] I have added or updated any relevant tests

### CODE REVIEW

**Comment by mcollina** [comment]
I would just go ahead and ship our own implementation instead of `subscription-transport-ws`.

File: examples/subscription/src/server/index.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
+const fastify = require('fastify')
+const GQL = require('fastify-gql')
+const mq = require('mqemitter')
+const emitter = mq()
```
[the remark this answers is not in the document]
Comment (frikille) [reply examples/subscription/src/server/index.js]: Will update the example using Redis

File: examples/subscription/src/server/index.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
+const fastify = require('fastify')
```
Comment (mcollina) [inline examples/subscription/src/server/index.js]: use strict is missing at the top of the file.

File: examples/subscription/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+const path = require('path')
```
Comment (mcollina) [inline examples/subscription/webpack.config.js]: use strict is missing

File: packages/graphql-hooks-subscription-client/test/index.test.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    server.close()
+    WS.clean()
+    server = new WS('ws://localhost:8888', { jsonProtocol: true })
+    server.send({ id: undefined, type: 'connection_ack' })
+
+    setTimeout(() => {
+      process.nextTick(() => {
+        server.send({ id: undefined, type: 'connection_ack' })
+      })
+    }, 1000)
```
Comment (mcollina) [inline packages/graphql-hooks-subscription-client/test/index.test.js]: I would prefer if we could rewrite this test to not use setTimeout.

File: packages/graphql-hooks/test/unit/useSubscription.test.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { renderHook } from '@testing-library/react-hooks'
+import { ClientContext, useSubscription, GraphQLClient } from '../../src'
+
+class MockSubscriptionClient {
```
Comment (mcollina) [inline packages/graphql-hooks/test/unit/useSubscription.test.js]: Why is this mock needed? Can you add some comments on why?

**Comment by simoneb** [comment]
@frikille thanks for your work on this, it's amazing! I skimmed through the code quickly and it all makes sense to me. The main question I have to start a discussion around this is, where does the code of `SubscriptionClient` come from? I suppose it comes from `subscription-transport-ws` and you copied its code here? 

If so, we may need to review how we want to move this forward, in the sense that we have a few options:

1) we reference that library
2) we copy the code
3) we implement our own

I think 1) doesn't adhere to the principle that's being applied throughout this library of avoiding third party dependencies. 2) would require making sure that we are allowed to copy the code and give proper attribution (haven't checked the licensing), and 3) I think is the preferred option, as well as the one requiring more effort of course.

**Comment by bmullan91** [comment]
@frikille thanks for your work here, as discussed the initial approach using `subscriptions-transport-ws` is ideal - we _really_ don't want to maintain our own flavour of that. I know you're busy right now, so I went ahead and cherry-picked the necessary commits over to #415 with a few tweaks.

**Pull request closed without merging**